pkg/cstor/migrate/pool.go

```diff
@@ -0,0 +1,324 @@
+			return nil, false, errors.Wrapf(err, "failed to get equivalent cspc for spc %s", spcName)
+		}
+		return nil, true, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	return spcObj, false, nil
+}
+
+// csptocspi migrates a CSP to CSPI based on hostname
+func csptocspi(
+	cspiObj *apis.CStorPoolInstance,
+	cspcObj *apis.CStorPoolCluster,
+	openebsNamespace string,
+) error {
+	hostnameLabel := string(apis.HostNameCPK) + "=" + cspiObj.Labels[string(apis.HostNameCPK)]
+	spcLabel := string(apis.StoragePoolClaimCPK) + "=" + cspcObj.Name
+	cspLabel := hostnameLabel + "," + spcLabel
+	var err1 error
+	cspObj, err := getCSP(cspLabel)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Migrating csp %s to cspi %s", cspiObj.Name, cspObj.Name)
+	err = scaleDownDeployment(cspObj, openebsNamespace)
+	if err != nil {
+		return err
+	}
+	// once the old pool pod is scaled down and bdcs are patched
+	// bring up the cspi pod so that the old pool can be renamed and imported.
+	cspiObj.Annotations[string(apis.OldPoolName)] = "cstor-" + string(cspObj.UID)
+	delete(cspiObj.Annotations, string(apis.OpenEBSDisableReconcileKey))
+	cspiObj, err = cspi.NewKubeClient().
+		WithNamespace(openebsNamespace).
+		Update(cspiObj)
+	if err != nil {
+		return err
+	}
+	err = retry.
+		Times(60).
+		Wait(5 * time.Second).
+		Try(func(attempt uint) error {
+			cspiObj, err1 = cspi.NewKubeClient().
+				WithNamespace(openebsNamespace).
+				Get(cspiObj.Name, metav1.GetOptions{})
+			if err1 != nil {
+				return err1
+			}
+			if cspiObj.Status.Phase != "ONLINE" {
+				return errors.Errorf("failed to verify cspi phase expected: Healthy got: %s",
+					cspiObj.Status.Phase)
+			}
+			return nil
+		})
+	if err != nil {
+		return err
+	}
+	err = updateCVRsLabels(cspObj.Name, openebsNamespace, cspiObj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// get csp for cspi on the basis of cspLabel, which is the combination of
+// hostname label on which cspi came up and the spc label.
+func getCSP(cspLabel string) (*apis.CStorPool, error) {
+	cspClient := csp.KubeClient()
+	cspList, err := cspClient.List(metav1.ListOptions{
+		LabelSelector: cspLabel,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(cspList.Items) != 1 {
+		return nil, fmt.Errorf("Invalid number of pools on one node: %d", len(cspList.Items))
+	}
+	cspObj := cspList.Items[0]
+	return &cspObj, nil
+}
+
+// The old pool pod should be scaled down before the new cspi pod comes up
+// to avoid importing the pool at two places at the same time.
+func scaleDownDeployment(cspObj *apis.CStorPool, openebsNamespace string) error {
+	klog.Infof("Scaling down deployemnt %s", cspObj.Name)
+	cspPod, err := pod.NewKubeClient().
+		WithNamespace(openebsNamespace).List(
+		metav1.ListOptions{
+			LabelSelector: "openebs.io/cstor-pool=" + cspObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+	if len(cspPod.Items) > 0 {
+		_, err = deploy.NewKubeClient().WithNamespace(openebsNamespace).
+			Patch(
+				cspObj.Name,
```

Comment:
Good to not assume the name of the CSP and deployment name as same. Can we get deployment name from pod owner reference?